presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java

```diff
@@ -129,6 +131,65 @@ public void testHistoryBasedStatsCalculator()
                 anyTree(node(AggregationNode.class, node(ExchangeNode.class, anyTree(any()))).withOutputRowCount(3).withOutputSize(54)));
     }
 
+    @Test
+    public void testHistoryBasedStatsCalculatorEnforceTimeOut()
+    {
+        Session session = Session.builder(createSession())
+                .setSystemProperty(ENFORCE_HISTORY_BASED_OPTIMIZER_REGISTRATION_TIMEOUT, "true")
+                .build();
+        Session alwaysTimeoutSession = Session.builder(createSession())
+                .setSystemProperty(ENFORCE_HISTORY_BASED_OPTIMIZER_REGISTRATION_TIMEOUT, "true")
+                .setSystemProperty(HISTORY_BASED_OPTIMIZER_TIMEOUT_LIMIT, "0ms")
+                .build();
+        // CBO Statistics
+        assertPlan(
+                session,
+                "SELECT * FROM nation where substr(name, 1, 1) = 'A'",
+                anyTree(node(FilterNode.class, any()).withOutputRowCount(Double.NaN)));
+
+        // HBO Statistics
```

Comment:
why is this using hbo statistics? Shouldn't registration fail due to exceeding the time limit?

Reply:
This is to test the case where HBO statistics are written successfully (with large timeout limit), and later it tests that read of HBO statistics either succeed (if timeout limit is large) or fail (if timeout limit is 0).

I just added a few new test cases too, which is HBO statistics write failed (with 0 timeout limit), and no HBO statistics will be read.